### ISSUE TITLE
vcvttss2usi

### ISSUE THREAD

**Issue by halx99** [simdsoft/x-studio#676]
```cpp
// original function name at ms vc++ stl is: _Min_load_factor_buckets, called by unordered_map internally.
std::size_t test_vcvttss2usi(const std::size_t _For_size)
{
    // returns the minimum number of buckets necessary for the elements in _List
    float fval = _CSTD ceilf(_For_size / 1.0f);

    /* The follow cast will lead crash
     __ftoui3: will crash at:
      OS: Windows Server 2008 R2 Enterprise
      CPU: Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz 2.50GHz
    */
    return static_cast<std::size_t>(fval);
}
```

**Comment by halx99** [comment]
__ftoui3

**Comment by halx99** [comment]
Visual Studio 2019 16.3.6
cl.exe: ```19.23.28106.4```
lib.exe & link.exe: ```14.23.28106.4```
Target platform: 
   OS: Windows Server 2008 R2 Enterprise
   CPU: Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz 2.50GHz

![image](https://user-images.githubusercontent.com/6977319/70633360-6df3e400-1c6b-11ea-9a9b-18a3fe3de09a.png)

**Comment by halx99** [comment]
https://devblogs.microsoft.com/cppblog/side-by-side-minor-version-msvc-toolsets-in-visual-studio-2019/

**Comment by halx99** [comment]
Solution: switch msvc tools version from ```14.23.28105``` to ```14.22.27905```
cl.exe version: ```19.22.27905```
lib.exe & link.exe version: ```14.22.27905.0```

**Comment by axhf** [comment]
Could you help to test and confirm whether [KB4494451](https://support.microsoft.com/en-us/help/4494451/kb4494451-intel-microcode-updates) solves this issue or not?

**Comment by halx99** [comment]
Thanks, I will try later

**Comment by halx99** [comment]
But I realized the patch is for ```Windows Server 2016 Version 1803Windows 10, version 1803```, but my server OS is Windows Server 2008

**Comment by axhf** [comment]
With the same hardware (Xeon Gold 5122, microcode version 0x200004D):
1. the issue appears on Windows Server 2012R2, with a minimum updates (for Universal CRT and .Net Framework)
2. no such issue on Windows Server 2016, with the latest updates

I was suspecting it is an issue with CPU microcode, when I did not have a test on the second server, but rather on a server with Xeon Gold 5222, microcode version 0x5000021 (the previously mentioned KB).

I'm still suspecting it is an issue with CPU microcode: maybe included in Windows Server 2016 or certain KB. But, I cannot do the tests by changing anything on those servers right now.

**Comment by halx99** [comment]
Thanks for your feedback, currently my solution make my software with max compatible for thus hardwares.

**Comment by halx99** [comment]
Visual Studio 2019 16.4.0
cl.exer: 19.24.28314.0

generated microcode:
![image](https://user-images.githubusercontent.com/6977319/70371501-acf20480-190e-11ea-9f09-af8d5f34600f.png)

The different with 19.23 is: 
```asm
cmp         dword ptr [__isa_available (0F9C520h)],6
 jl          _ftoui3_default (0F92A40h) 
```

but actualy, my computer windows 10, intel i7 6700HQ support vcvttss2usi

**Comment by axhf** [comment]
I believe it is [fixed in 16.4](https://github.com/dotnet/wpf/issues/2259#issuecomment-562695956).

I don't see I can verify it on the previous hardware.
But, the [sample code](https://github.com/dotnet/wpf/issues/2259#issuecomment-562716372) is almost the same as mine (below) and similar to yours.

```c++
int main()
{
	volatile float a = 1.0f;
	return (int)(unsigned int)(a);
}
```

**Comment by halx99** [comment]
yes, it will fix the crash issue, but the conditon of generated microcode maybe incorrect, because, my computer, intel i7 6700HQ, win10 support vcvttss2usi, but the actual code can't execute vcvttss2usi on my computer

**Comment by halx99** [comment]
VS 2019 16.4 14.22 minor version upgrade from 27905 to 27906
![image](https://user-images.githubusercontent.com/6977319/70459643-948d1000-1aef-11ea-91c6-deaba880dab6.png)

16.3 14.22 minor version is:
![image](https://user-images.githubusercontent.com/6977319/70489309-5d8e1d00-1b36-11ea-9e81-d1b56dd33f01.png)

**Comment by halx99** [comment]
microcode generated by follow vctools revisions:
vctools 14.22: cvttss2si
vctools 14.23: vcvttss2usi
vctools 14.24: cvttss2si

**Comment by halx99** [comment]
Actually: since VS 2019 16.3.10
cl.exe: 19.23.28107.0
link.exe: 14.23.28107.0

crash is fixed, and generated microcode is same as 16.4:
![image](https://user-images.githubusercontent.com/6977319/70495578-e9a94000-1b48-11ea-88b1-e7d18fb99de9.png)

**Comment by halx99** [comment]
The VS 2019 16.3.6 generated microcode is:

![image](https://user-images.githubusercontent.com/6977319/70636330-4e12ef00-1c70-11ea-801d-b16e58ed929b.png)


cmp         dword ptr ds:[733529A0h],6  

The memory is 6, so crashed with 0xC000001D: Illegal Instruction.
